Add tram and subway

DO3B opened the pull request:
This is a draft PR  (https://github.com/e-mission/e-mission-docs/issues/428) which will add tram and subway to the GIS branch of e-mission-server.

AmplabJenkins commented:
Can one of the admins verify this patch?

shankari commented on emission/analysis/classification/inference/mode/rule_engine.py:
```diff
@@ -187,8 +187,10 @@ def collapse_modes(section_entry, modes):
-        'light_rail', 'subway', 'monorail', 'tram', 'aerialway', ]
+    # train_mode_list = ['funicular', 'miniature', 'rail', 'railway',
+    #     'light_rail', 'subway', 'monorail', 'tram', 'aerialway', ]
+    train_mode_list = ['funicular', 'miniature', 'rail', 'railway', 'light_rail', 'monorail', 
```
just to confirm my own understanding here: with this change, you are removing `subway` and `tram` from the list of modes that we should collapse to `train`. What about `track`, the mode that originally started this entire discussion? https://github.com/e-mission/e-mission-docs/issues/426

DO3B replied:
Well, I didn't encounter this issue but I should definitely put "tracks" back on this list, just in case.

shankari commented on emission/net/ext_service/transit_matching/match_stops.py:
```diff
@@ -96,17 +97,28 @@ def get_predicted_transit_mode(start_stops, end_stops):
-        return ["TRAIN"]
-
+    p_start_train = [extract_railway_modes(s.tags) for s in start_stops]
+    p_start_train = Counter(itertools.chain.from_iterable(set(i) for i in p_start_train))
```
Again, for my own understanding, we have a set of tags for each potential start stop, and the `from_iterable` converts them into one giant list which you then use `Counter` over. I should really use `itertools` more instead of manually invoking list comprehensions all the time :)

shankari commented on emission/net/ext_service/transit_matching/match_stops.py:
```diff
@@ -96,17 +97,28 @@ def get_predicted_transit_mode(start_stops, end_stops):
+    if len(p_start_train) > 0 and len(p_end_train) > 0:
+        p_start_max = p_start_train.most_common(1)[0][0]
+        p_end_max = p_end_train.most_common(1)[0][0]
+        if p_start_max is not None and p_start_max == p_end_max:
```
Are you sure that building a check from the `most_common` is the best strategy? What if there is an overlapping mode, but it is not the most common? Would a set intersection work better? I know that @PatGendre said that he has taken tram trips, have you tried running the algorithm for any of them to see how well it works?

DO3B replied:
When I tested on the data that I had (some trips in Berlin), I found that most of the time the `most_common` was the right mode but after tested it recently on new data in Marseille, it was not the case so I will try to use a set intersection.

DO3B commented:
I mostly tested it on some data in Berlin where the most common strategy was working but I tested it on those data recently (I used subway and bus during a day) and it worked one time for the subway but I took it at least 3 times during this day. But those data are not working anymore...

```
2019-08-05 10:20:40,847:DEBUG:139776991606528:first element in section_location_array = Entry({'_id': ObjectId('5d47e6ba4e276d098ba1fffc'), 'user_id': UUID('90b6ed83-f2ab-4c0b-b570-b8dc07e1e325'), 'metadata': {'key': 'analysis/recreated_location', 'platform': 'server', 'write_ts': 1564993210.315876, 'time_zone': 'America/Los_Angeles', 'write_local_dt': {'year': 2019, 'month': 8, 'day': 5, 'hour': 1, 'minute': 20, 'second': 10, 'weekday': 0, 'timezone': 'America/Los_Angeles'}, 'write_fmt_time': '2019-08-05T01:20:10.315876-07:00'}, 'data': {'latitude': 43.2911525, 'longitude': 5.4003441, 'loc': {'type': 'Point', 'coordinates': [5.4003441, 43.2911525]}, 'ts': 1564227153.778, 'local_dt': {'year': 2019, 'month': 7, 'day': 27, 'hour': 13, 'minute': 32, 'second': 33, 'weekday': 5, 'timezone': 'Europe/Paris'}, 'fmt_time': '2019-07-27T13:32:33.778000+02:00', 'altitude': 61.55169677734375, 'distance': 0.24355621424840848, 'speed': 0.04735978469907235, 'heading': -48.87999607846491, 'idx': 29, 'mode': 2, 'section': ObjectId('5d47e6ba4e276d098ba1ffde')}})
2019-08-05 10:20:40,848:INFO:139776991606528:section_location_array[-1].data.loc {"coordinates": [5.4003441, 43.2911525], "type": "Point"} != section.data.end_loc {"coordinates": [5.3922696, 43.27173], "type": "Point"} even after df.ts fix, filling gap
2019-08-05 10:20:40,848:ERROR:139776991606528:Found error  while processing trip Entry({'_id': ObjectId('5d47e6ba4e276d098ba1ffdc'), 'user_id': UUID('90b6ed83-f2ab-4c0b-b570-b8dc07e1e325'), 'metadata': {'key': 'analysis/cleaned_trip', 'platform': 'server', 'write_ts': 1564993210.0501163, 'time_zone': 'America/Los_Angeles', 'write_local_dt': {'year': 2019, 'month': 8, 'day': 5, 'hour': 1, 'minute': 20, 'second': 10, 'weekday': 0, 'timezone': 'America/Los_Angeles'}, 'write_fmt_time': '2019-08-05T01:20:10.050116-07:00'}, 'data': {'source': 'DwellSegmentationTimeFilter', 'end_ts': 1564228358.01, 'end_local_dt': {'year': 2019, 'month': 7, 'day': 27, 'hour': 13, 'minute': 52, 'second': 38, 'weekday': 5, 'timezone': 'Europe/Paris'}, 'end_fmt_time': '2019-07-27T13:52:38.010000+02:00', 'end_loc': {'type': 'Point', 'coordinates': [5.39217, 43.2706194]}, 'raw_trip': ObjectId('5d47e6b44e276d098ba1feef'), 'start_ts': 1564226308.6353202, 'start_local_dt': {'year': 2019, 'month': 7, 'day': 27, 'hour': 13, 'minute': 18, 'second': 28, 'weekday': 5, 'timezone': 'Europe/Paris'}, 'start_fmt_time': '2019-07-27T13:18:28.635320+02:00', 'start_loc': {'type': 'Point', 'coordinates': [5.403022, 43.2919448]}, 'duration': 2049.3746798038483, 'distance': 2740.8649781758345, 'start_place': ObjectId('5d47e6be4e276d098ba2010b'), 'end_place': ObjectId('5d47e6be4e276d098ba2010c')}})
Traceback (most recent call last):
  File "/var/emission/e-mission-server/emission/analysis/plotting/geojson/geojson_feature_converter.py", line 284, in get_geojson_for_timeline
    trip_geojson = trip_to_geojson(trip, tl)
  File "/var/emission/e-mission-server/emission/analysis/plotting/geojson/geojson_feature_converter.py", line 243, in trip_to_geojson
    section_gj = section_to_geojson(section, tl)
  File "/var/emission/e-mission-server/emission/analysis/plotting/geojson/geojson_feature_converter.py", line 134, in section_to_geojson
    assert(False)
AssertionError
2019-08-05 10:20:40,859:ERROR:139776991606528:Storing views to cache failed for user 90b6ed83-f2ab-4c0b-b570-b8dc07e1e325
Traceback (most recent call last):
  File "/var/emission/e-mission-server/emission/net/usercache/builtin_usercache_handler.py", line 109, in storeViewsToCache
    self.storeTimelineToCache(time_query)
  File "/var/emission/e-mission-server/emission/net/usercache/builtin_usercache_handler.py", line 180, in storeTimelineToCache
    trip_gj_list = self.get_trip_list_for_seven_days(start_ts)
  File "/var/emission/e-mission-server/emission/net/usercache/builtin_usercache_handler.py", line 269, in get_trip_list_for_seven_days
    trip_gj_list = gfc.get_geojson_for_ts(self.user_id, seventy_two_hours_ago_ts, start_ts)
  File "/var/emission/e-mission-server/emission/analysis/plotting/geojson/geojson_feature_converter.py", line 268, in get_geojson_for_ts
    return get_geojson_for_timeline(user_id, tl)
  File "/var/emission/e-mission-server/emission/analysis/plotting/geojson/geojson_feature_converter.py", line 289, in get_geojson_for_timeline
    raise e
  File "/var/emission/e-mission-server/emission/analysis/plotting/geojson/geojson_feature_converter.py", line 284, in get_geojson_for_timeline
    trip_geojson = trip_to_geojson(trip, tl)
  File "/var/emission/e-mission-server/emission/analysis/plotting/geojson/geojson_feature_converter.py", line 243, in trip_to_geojson
    section_gj = section_to_geojson(section, tl)
  File "/var/emission/e-mission-server/emission/analysis/plotting/geojson/geojson_feature_converter.py", line 134, in section_to_geojson
AssertionError
2019-08-05 10:20:40,859:ERROR:139776991606528:Storing views to cache failed for user 90b6ed83-f2ab-4c0b-b570-b8dc07e1e325
Traceback (most recent call last):
  File "/var/emission/e-mission-server/emission/net/usercache/builtin_usercache_handler.py", line 109, in storeViewsToCache
    self.storeTimelineToCache(time_query)
  File "/var/emission/e-mission-server/emission/net/usercache/builtin_usercache_handler.py", line 180, in storeTimelineToCache
    trip_gj_list = self.get_trip_list_for_seven_days(start_ts)
  File "/var/emission/e-mission-server/emission/net/usercache/builtin_usercache_handler.py", line 269, in get_trip_list_for_seven_days
    trip_gj_list = gfc.get_geojson_for_ts(self.user_id, seventy_two_hours_ago_ts, start_ts)
  File "/var/emission/e-mission-server/emission/analysis/plotting/geojson/geojson_feature_converter.py", line 268, in get_geojson_for_ts
    return get_geojson_for_timeline(user_id, tl)
  File "/var/emission/e-mission-server/emission/analysis/plotting/geojson/geojson_feature_converter.py", line 289, in get_geojson_for_timeline
    raise e
  File "/var/emission/e-mission-server/emission/analysis/plotting/geojson/geojson_feature_converter.py", line 284, in get_geojson_for_timeline
    trip_geojson = trip_to_geojson(trip, tl)
  File "/var/emission/e-mission-server/emission/analysis/plotting/geojson/geojson_feature_converter.py", line 243, in trip_to_geojson
    section_gj = section_to_geojson(section, tl)
  File "/var/emission/e-mission-server/emission/analysis/plotting/geojson/geojson_feature_converter.py", line 134, in section_to_geojson
    assert(False)
AssertionError
```
[subway_test.txt](https://github.com/e-mission/e-mission-server/files/3466575/subway_test.txt)

shankari commented:
@DO3B the assertion error is from 

```
        if not ecc.compare_rounded_arrays(section.data.end_loc.coordinates,
                                      section_location_entries[-1].data.loc.coordinates,
                                      digits=4):
            logging.info("section_location_array[-1].data.loc %s != section.data.end_loc %s even after df.ts fix, filling gap" % \
                    (section_location_entries[-1].data.loc, section.data.end_loc))
            assert(False)
            last_loc_doc = ts.get_entry_at_ts("background/filtered_location", "data.ts", section.data.end_ts)
            if last_loc_doc is None:
                logging.warning("can't find entry to patch gap, leaving gap")
            else:
                last_loc_entry = ecwe.Entry(last_loc_doc)
                logging.debug("Adding new entry %s to fill the end point gap between %s and %s"
                   % (last_loc_entry.data.loc, section_location_entries[-1].data.loc,
                        section.data.end_loc))
                section_location_entries.append(last_loc_entry)
```

It should be fine to comment out that assert since we try to workaround it anyway.

shankari commented:
@DO3B 

> When I tested on the data that I had (some trips in Berlin), I found that most of the time the most_common was the right mode but after tested it recently on new data in Marseille, it was not the case so I will try to use a set intersection.

But the most common v/s intersection code only comes into play iff there is no relation for the trips. But at least in the US, train trips do have associated relations. Let me run your data and see what I can figure out...

shankari commented:
Looking at the logs for your data, I see:

We start the mode inference step

```
2019-08-05 12:48:06,583:INFO:4667995584:For stage PipelineStages.MODE_INFERENCE, start_ts is None
```

For this section

```
2019-08-05 12:48:06,592:DEBUG:4667995584:~~~~~~~~~~About to get predicted value for section 5d4887e9f6858f77c523d223 (2019-07-27T11:44:17.234083+02:00 -> 2019-07-27T12:12:17+02:00)~~~~~~~~~~
```

we have 75(!!) relations, some of which have nodes and some which don't.
After listing the relations and the nodes, we iterate over them to find potential matches and it essentially boils down to bus.

```
2019-08-05 12:48:24,273:DEBUG:4667995584:After iterating through routes, potential modes = [(6626048, 'bus')]
....
2019-08-05 12:48:24,569:DEBUG:4667995584:matching routes = [(6636620, 'LER 29'), (6636621, 'LER 29'), (6636622, 'LER 29'), (6636708, 'LER 29'), (6636709, 'LER 29'), (6636710, 'LER 29'), (6781532, 'LER 28'), (6781533, 'LER 28'), (6623330, '50'), (6623329, '50'), (6784526, 'LER 25'), (6784527, 'LER 25'), (7168335, 'LER 27'), (7168336, 'LER 27'), (9051016, 'LER 20'), (9051017, 'LER 20')]
2019-08-05 12:48:24,569:DEBUG:4667995584:len(start_routes) = 53, len(end_routes) = 388, len(rel_id_matches) = 16
...
2019-08-05 12:48:24,570:DEBUG:4667995584:Got predicted transit mode ['bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus']
2019-08-05 12:48:24,570:DEBUG:4667995584:train_mapped_modes = ['BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS']
2019-08-05 12:48:24,570:DEBUG:4667995584:unique_modes = ['BUS']
```

Note that we are able to use the route relations to find the match 

```
    logging.debug("len(start_routes) = %d, len(end_routes) = %d, len(rel_id_matches) = %d" %
        (len(p_start_routes), len(p_end_routes), len(rel_id_matches)))
    if len(rel_id_matches) > 0:
        return [rim.tags.route for rim in rel_id_matches]
```

Similarly checking the next section...

```
2019-08-05 12:48:24,574:DEBUG:4667995584:~~~~~~~~~~About to get predicted value for section 5d4887eaf6858f77c523d2d3 (2019-07-27T13:32:33.778000+02:00 -> 2019-07-27T13:49:55.500000+02:00)~~~~~~~~~~
2019-08-05 12:48:29,409:DEBUG:4667995584:matching routes = [(6146575, '72'), (6146575, '72'), (6146575, '72'), (6146575, '72'), (6195617, '72'), (6195617, '72'), (6195617, '72'), (6195617, '72')]
2019-08-05 12:48:29,409:DEBUG:4667995584:Got predicted transit mode ['bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus']
2019-08-05 12:48:29,409:DEBUG:4667995584:train_mapped_modes = ['BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS']
2019-08-05 12:48:29,409:DEBUG:4667995584:unique_modes = ['BUS']
```

And checking the next section

```
2019-08-05 12:48:29,411:DEBUG:4667995584:~~~~~~~~~~About to get predicted value for section 5d4887ebf6858f77c523d2dd (2019-07-27T13:59:49.831605+02:00 -> 2019-07-27T14:18:25.007000+02:00)~~~~~~~~~~
2019-08-05 12:48:34,098:DEBUG:4667995584:matching routes = [(2102777, '21J'), (2102777, '21J'), (2102777, '21J'), (2102777, '21J'), (2105624, '21J'), (2105624, '21J'), (2105624, '21J'), (2105624, '21J'), (1656744, 'B1'), (1656744, 'B1'), (1656744, 'B1'), (1656744, 'B1'), (6429595, '521'), (6429595, '521'), (6429595, '521'), (6429595, '521'), (1656745, 'B1'), (1656745, 'B1'), (1656745, 'B1'), (1656745, 'B1'), (6429594, '521'), (6429594, '521'), (6429594, '521'), (6429594, '521'), (1656744, 'B1'), (1656744, 'B1'), (1656744, 'B1'), (1656744, 'B1'), (6429595, '521'), (6429595, '521'), (6429595, '521'), (6429595, '521'), (2102777, '21J'), (2102777, '21J'), (2102777, '21J'), (2102777, '21J'), (2105624, '21J'), (2105624, '21J'), (2105624, '21J'), (2105624, '21J'), (1656745, 'B1'), (1656745, 'B1'), (1656745, 'B1'), (1656745, 'B1'), (6429594, '521'), (6429594, '521'), (6429594, '521'), (6429594, '521')]
2019-08-05 12:48:34,099:DEBUG:4667995584:len(start_routes) = 37, len(end_routes) = 32, len(rel_id_matches) = 48
2019-08-05 12:48:34,099:DEBUG:4667995584:Got predicted transit mode ['bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus']
2019-08-05 12:48:34,100:DEBUG:4667995584:train_mapped_modes = ['BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS']
2019-08-05 12:48:34,100:DEBUG:4667995584:unique_modes = ['BUS']
```

And checking the next section. In this one, we did not find any matching relations, but we did not find any matching stops either.

```
2019-08-05 12:48:34,102:DEBUG:4667995584:~~~~~~~~~~About to get predicted value for section 5d4887ebf6858f77c523d30c (2019-07-27T14:22:58.293000+02:00 -> 2019-07-27T14:47:29+02:00)~~~~~~~~~~
2019-08-05 12:48:39,872:DEBUG:4667995584:matching routes = []
2019-08-05 12:48:39,872:DEBUG:4667995584:len(start_routes) = 16, len(end_routes) = 4, len(rel_id_matches) = 0
2019-08-05 12:48:39,873:DEBUG:4667995584:len(start_train) = 4, len(end_train) = 2
2019-08-05 12:48:39,873:DEBUG:4667995584:is_true([False, False, False, False]) = False
2019-08-05 12:48:39,873:DEBUG:4667995584:4 start stops -> 2 start bus stops, 2 end stops -> 1 end bus stops
2019-08-05 12:48:39,874:DEBUG:4667995584:2 bus stops -> 0 bus stops with refs -> 0 routes
2019-08-05 12:48:39,874:DEBUG:4667995584:1 bus stops -> 0 bus stops with refs -> 0 routes
2019-08-05 12:48:39,874:DEBUG:4667995584:About to find matches in lists: [] 
 []
2019-08-05 12:48:39,874:DEBUG:4667995584:matching routes = []
2019-08-05 12:48:39,875:DEBUG:4667995584:len(start_bus_route_ref) = 0, len(end_bus_route_ref) = 0, len(rel_id_matches) = 0
2019-08-05 12:48:39,875:DEBUG:4667995584:len(start_bus) = 4, len(end_bus) = 2
2019-08-05 12:48:39,875:DEBUG:4667995584:is_true([False, True, False, True]) = True
2019-08-05 12:48:39,875:DEBUG:4667995584:is_true([True, False]) = True
2019-08-05 12:48:39,876:DEBUG:4667995584:Both start and end are bus stops, validating...
2019-08-05 12:48:39,876:DEBUG:4667995584:Both side are dense, invalid bus stop
2019-08-05 12:48:39,876:DEBUG:4667995584:Got predicted transit mode None
2019-08-05 12:48:39,877:DEBUG:4667995584:no luck finding a match, stay UNKNOWN
```

Finishing with the last two, both of them use the relations to map to `BUS`

```
2019-08-05 12:48:39,882:DEBUG:4667995584:~~~~~~~~~~About to get predicted value for section 5d4887ecf6858f77c523d375 (2019-07-27T19:47:16.099601+02:00 -> 2019-07-27T19:50:49.011000+02:00)~~~~~~~~~~
2019-08-05 12:48:47,205:DEBUG:4667995584:matching routes = [(1656744, 'B1'), (1656744, 'B1'), (6429595, '521'), (6429595, '521'), (1656745, 'B1'), (1656745, 'B1'), (1656744, 'B1'), (1656744, 'B1'), (1656745, 'B1'), (1656745, 'B1'), (6429595, '521'), (6429595, '521'), (1656745, 'B1'), (1656745, 'B1'), (1656744, 'B1'), (1656744, 'B1'), (6429595, '521'), (6429595, '521'), (1656745, 'B1'), (1656745, 'B1'), (1656744, 'B1'), (1656744, 'B1'), (6429595, '521'), (6429595, '521')]
2019-08-05 12:48:47,205:DEBUG:4667995584:len(start_routes) = 30, len(end_routes) = 14, len(rel_id_matches) = 24
2019-08-05 12:48:47,205:DEBUG:4667995584:Got predicted transit mode ['bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus']
2019-08-05 12:48:47,205:DEBUG:4667995584:train_mapped_modes = ['BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS']
2019-08-05 12:48:47,205:DEBUG:4667995584:unique_modes = ['BUS']
```

```
2019-08-05 12:48:47,206:DEBUG:4667995584:~~~~~~~~~~About to get predicted value for section 5d4887edf6858f77c523d38d (2019-07-27T19:58:20.102000+02:00 -> 2019-07-27T20:15:57.815000+02:00)~~~~~~~~~~
2019-08-05 12:48:59,622:DEBUG:4667995584:matching routes = [(6429594, '521'), (6429594, '521'), (6429595, '521'), (6429595, '521'), (6429595, '521'), (6429595, '521'), (6429595, '521'), (6429595, '521')]
2019-08-05 12:48:59,622:DEBUG:4667995584:len(start_routes) = 39, len(end_routes) = 48, len(rel_id_matches) = 8
2019-08-05 12:48:59,623:DEBUG:4667995584:Got predicted transit mode ['bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus', 'bus']
2019-08-05 12:48:59,623:DEBUG:4667995584:train_mapped_modes = ['BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS', 'BUS']
2019-08-05 12:48:59,623:DEBUG:4667995584:unique_modes = ['BUS']
2019-08-05 12:48:59,623:DEBUG:4667995584:predicted transit mode is not None, UNKNOWN -> BUS
```

Basically, it looks like the relation check is the one that is used in almost every case; we don't fall back to the stop check except in one case where (atleast without your changes) we did not find a match anyway

shankari commented:
I didn't say that my change fixes this.
Dunno why github closed it.

AmplabJenkins commented:
Can one of the admins verify this patch?

shankari commented:
@DO3B any updates on the subway detection with the assert fix? Which of those trips was supposed to be a subway? Given almost all of them returned after the relation check, do you really need any changes to the stop check?

DO3B commented:
In most of the case, there is indeed a relation check, but sometimes, we reach the case where we need to extract the stop tags.

```
2019-08-07 13:20:27,394:DEBUG:139895478814464:~~~~~~~~~~About to get predicted value for section 5d4ab2df4e276d1178a572e2 (2019-07-27T19:58:20.102000+02:00 -> 2019-07-27T20:15:57.815000+02:00)~~~~~~~~~~
2019-08-07 13:21:27,510:DEBUG:139895478814464:matching routes = []
2019-08-07 13:21:27,510:DEBUG:139895478814464:len(start_routes) = 18, len(end_routes) = 33, len(rel_id_matches) = 0
2019-08-07 13:21:27,511:DEBUG:139895478814464:After extracting data from tags, potential modes = ['SUBWAY']
2019-08-07 13:21:27,511:DEBUG:139895478814464:After extracting data from tags, potential modes = []
2019-08-07 13:21:27,511:DEBUG:139895478814464:After extracting data from tags, potential modes = []
2019-08-07 13:21:27,511:DEBUG:139895478814464:After extracting data from tags, potential modes = []
2019-08-07 13:21:27,512:DEBUG:139895478814464:After extracting data from tags, potential modes = []
2019-08-07 13:21:27,512:DEBUG:139895478814464:After extracting data from tags, potential modes = []
2019-08-07 13:21:27,512:DEBUG:139895478814464:After extracting data from tags, potential modes = []
2019-08-07 13:21:27,512:DEBUG:139895478814464:After extracting data from tags, potential modes = ['SUBWAY']
2019-08-07 13:21:27,512:DEBUG:139895478814464:After extracting data from tags, potential modes = ['SUBWAY']
2019-08-07 13:21:27,513:DEBUG:139895478814464:After extracting data from tags, potential modes = ['SUBWAY']
2019-08-07 13:21:27,513:DEBUG:139895478814464:After extracting data from tags, potential modes = []
2019-08-07 13:21:27,513:DEBUG:139895478814464:After extracting data from tags, potential modes = []
2019-08-07 13:21:27,513:DEBUG:139895478814464:After extracting data from tags, potential modes = []
2019-08-07 13:21:27,514:DEBUG:139895478814464:After extracting data from tags, potential modes = []
2019-08-07 13:21:27,514:DEBUG:139895478814464:After extracting data from tags, potential modes = ['SUBWAY']
2019-08-07 13:21:27,514:DEBUG:139895478814464:After extracting data from tags, potential modes = ['SUBWAY']
2019-08-07 13:21:27,514:DEBUG:139895478814464:len(start_train) = 1, len(end_train) = 1
2019-08-07 13:21:27,514:DEBUG:139895478814464:Start and end are both SUBWAY, returning SUBWAY
2019-08-07 13:21:27,515:DEBUG:139895478814464:Got predicted transit mode ['SUBWAY']
2019-08-07 13:21:27,515:DEBUG:139895478814464:train_mapped_modes = ['SUBWAY']
2019-08-07 13:21:27,515:DEBUG:139895478814464:predicted transit mode is not None, UNKNOWN -> SUBWAY
```

With your recent update, one of the trip which was supposed to be a subway is now properly detected and was before detected as a bus trip (the section above).

shankari commented:
Hm, this is pretty bizarre since all of my updates only commented out some asserts and should not have changed any actual implementations. When I look at my logs for that section, I see a bunch of matching bus routes.

```
2019-08-05 12:48:47,206:DEBUG:4667995584:~~~~~~~~~~About to get predicted value for section 5d4887edf6858f77c523d38d (2019-07-27T19:58:20.102000+02:00 -> 2019-07-27T20:15:57.815000+02:00)~~~~~~~~~~
2019-08-05 12:48:55,804:DEBUG:4667995584:Found 23 relations with ids [1419198, 1642659, 1642660, 1656425, 1656744, 1656745, 1771619, 1831119, 1831120, 2102777, 2105624, 2202168, 2202169, 2222449, 2222450, 2222451, 2222452, 6146575, 6195617, 6274532, 6274533, 6429594, 6429595]
2019-08-05 12:48:59,386:DEBUG:4667995584:Found 53 relations with ids [1656424, 1656680, 1656681, 1656709, 1691204, 1691205, 1694218, 1763272, 1768990, 1768991, 1770801, 1881786, 1881787, 2133002, 2133003, 2133004, 2868847, 2868848, 5386529, 5392366, 5441743, 6160562, 6274530, 6274531, 6354090, 6354120, 6354121, 6397720, 6397721, 6398462, 6398463, 6398464, 6398465, 6408054, 6408055, 6429594, 6429595, 6447743, 6447744, 6447844, 6447845, 6458213, 6458214, 6461475, 6461476, 6487092, 6487093, 6491638, 6491639, 6492000, 6492001, 6503794, 6503795]
```

There are two subway route relations, but they are not the same

```
2019-08-05 12:48:59,593:DEBUG:4667995584:After iterating through routes, potential modes = [(6274532, 'subway')]
2019-08-05 12:48:59,596:DEBUG:4667995584:After iterating through routes, potential modes = [(6274533, 'subway')]
```

So they don't show up in the matches

```
2019-08-05 12:48:59,622:DEBUG:4667995584:matching routes = [(6429594, '521'), (6429594, '521'), (6429595, '521'), (6429595, '521'), (6429595, '521'), (6429595, '521'), (6429595, '521'), (6429595, '521')]
```

Comparing with your logs, I see that I had way more start and end routes than you did.

```
2019-08-05 12:48:59,622:DEBUG:4667995584:len(start_routes) = 39, len(end_routes) = 48, len(rel_id_matches) = 8
2019-08-07 13:21:27,510:DEBUG:139895478814464:len(start_routes) = 18, len(end_routes) = 33, len(rel_id_matches) = 0
```

which could be the reason, but I don't understand why overpass is returning different results to you and me. Did you change the start and stop radius by any chance?

shankari commented:
I just re-ran and got identical results. Do you have any changes to the overpass API call that are causing the results to be different?

```
2019-08-07 12:13:54,705:DEBUG:4593313216:~~~~~~~~~~About to get predicted value for section 5d4b2261f6858f4e572c2faa (2019-07-27T19:58:20.102000+02:00 -> 2019-07-27T20:15:57.815000+02:00)~~~~~~~~~~
2019-08-07 12:14:09,413:DEBUG:4593313216:Found 23 relations with ids [1419198, 1642659, 1642660, 1656425, 1656744, 1656745, 1771619, 1831119, 1831120, 2102777, 2105624, 2202168, 2202169, 2222449, 2222450, 2222451, 2222452, 6146575, 6195617, 6274532, 6274533, 6429594, 6429595]
2019-08-07 12:14:42,052:DEBUG:4593313216:Found 53 relations with ids [1656424, 1656680, 1656681, 1656709, 1691204, 1691205, 1694218, 1763272, 1768990, 1768991, 1770801, 1881786, 1881787, 2133002, 2133003, 2133004, 2868847, 2868848, 5386529, 5392366, 5441743, 6160562, 6274530, 6274531, 6354090, 6354120, 6354121, 6397720, 6397721, 6398462, 6398463, 6398464, 6398465, 6408054, 6408055, 6429594, 6429595, 6447743, 6447744, 6447844, 6447845, 6458213, 6458214, 6461475, 6461476, 6487092, 6487093, 6491638, 6491639, 6492000, 6492001, 6503794, 6503795]
```

There are two subway route relations, but they are not the same

```
2019-08-07 12:14:42,302:DEBUG:4593313216:After iterating through routes, potential modes = [(6274532, 'subway')]
2019-08-07 12:14:42,304:DEBUG:4593313216:After iterating through routes, potential modes = [(6274533, 'subway')]
```

So they don't show up in the matches

```
2019-08-07 12:14:42,329:DEBUG:4593313216:matching routes = [(6429594, '521'), (6429594, '521'), (6429595, '521'), (6429595, '521'), (6429595, '521'), (6429595, '521'), (6429595, '521'), (6429595, '521')]
```

Comparing with your logs, I see that I had way more start and end routes than you did.

```
2019-08-07 12:14:42,329:DEBUG:4593313216:len(start_routes) = 39, len(end_routes) = 48, len(rel_id_matches) = 8
2019-08-07 13:21:27,510:DEBUG:139895478814464:len(start_routes) = 18, len(end_routes) = 33, len(rel_id_matches) = 0
```

DO3B commented:
Oh ! Sorry, I sent you the wrong logs. Those are the logs when I had set the radius to 50. As you noticed, since I had a lot of data, I tried to reduce the radius to not reach Overpass API limitation and have a better response time. However, I ran the pipeline later during the day and had the same result than you.

shankari commented:
Looking further into why the relations did not match, there are actually 4 subway matches

```
2019-08-07 12:14:42,302:DEBUG:4593313216:After iterating through routes, potential modes = [(6274532, 'subway')]
2019-08-07 12:14:42,304:DEBUG:4593313216:After iterating through routes, potential modes = [(6274533, 'subway')]
```

```
2019-08-07 12:14:42,304:DEBUG:4593313216:After iterating through routes, potential modes = [(6274531, 'subway')]
2019-08-07 12:14:42,306:DEBUG:4593313216:After iterating through routes, potential modes = [(6274530, 'subway')]
```

These relations form two pairs
6274532 and 6274533 are the two directions (Bougainville -> Dromel, Dromel -> Bougainville) of the parent relation https://www.openstreetmap.org/relation/165618 which is the M2 line

6274530 and 6274531 are the two directions (La Fourragère => La Rose, La Rose => La Fourragère) of the parent relation https://www.openstreetmap.org/relation/165613 which is the M1 line

```
2019-08-07 12:14:42,307:DEBUG:4593313216:About to find matches in lists: [6274532, ...,  6274533, 6146575] 
[6274531, ...,  6274530, 1881786]
```

So it looks like you started on the M2 line, transferred and then got off from the M1 line. It seems like we should be able to detect this from OSM, but will have to poke around the data to figure out how.

shankari commented:
So I notice that all the metro relations have `network=RTM`. So one workaround may be to change the relation matching code to not just check the relation id, but also, if the two modes are both train related, to see whether they are part of the same network?

I don't think this will work for bus because of the density of bus routes/stops, but it should work for train modes, which are a lot less dense.

DO3B commented:
I noticed that we were not using the `overpass_server.json.sample` and added a quick fix (https://github.com/e-mission/e-mission-server/pull/709/commits/8664a77c987be16e8e025f323ff127375f357354). I also changed the "Retrying after 25 seconds sleep" by a `GET` to `api/status` to know when the next slot is available (if the two slots are used).

PatGendre commented:
Thank you @DO3B ! This may solve largely the api errors we had when too many calls were submitted, e.g. when resetting/relaunching the pipeline.

shankari commented:
@DO3B that's great. Any thoughts on the subway relation matching that I proposed?

DO3B commented:
I can try it tomorrow since most of the public transport in a city belong to the same network (in Europe at least) that should improve our results (and detect when someone transit).

shankari commented:
@DO3B great! Can you also confirm that this is what happened?

> So it looks like you started on the M2 line, transferred and then got off from the M1 line. It seems like we should be able to detect this from OSM, but will have to poke around the data to figure out how.

DO3B commented:
@shankari  Yup, but I took the M1 first and then the M2.

DO3B commented:
I also had to check if the "network" tags exists because I had a surprise in Berlin where only one train relation has no network tags :(

shankari commented:
> I also had to check if the "network" tags exists because I had a surprise in Berlin where only one train relation has no network tags :(

Yes, the more I work with OSM data, the more I find such corner-cases.

Do you want to change the stop check to intersection of sets instead of the most common entry? I am fine with deferring it, since we typically don't fall back to it anyway. You also need to make the UI changes before you are done...

DO3B commented:
I already did the UI changes, but I added FontAwesome to have the new icons (tram and subway) https://github.com/fabmob/e-mission-phone-fabmob/issues/21#issuecomment-514617164. I just have to PR them :) 

I will do the intersection today.

shankari commented:
LGTM!
